javaslang/src/main/java/javaslang/collection/LeafType.java

```diff
@@ -5,99 +5,238 @@
+    static final ObjectArray OBJECT_TYPE = new ObjectArray();
+
+    @Override
+    boolean isPrimitive() { return false; }
+
+    @Override
+    Class<Object> type() { return Object.class; }
+
+    static final Object[] EMPTY = new Object[0];
+
+    @Override
+    Object[] empty() { return EMPTY; }
+
+    private static Object[] cast(Object array) { return (array == null) ? EMPTY : (Object[]) array; }
+
+    @Override
+    int lengthOf(Object array) { return cast(array).length; }
+
+    @Override
+    Object getAt(Object array, int index) { return cast(array)[index]; }
+
+    @Override
+    void setAt(Object array, int index, Object value) { cast(array)[index] = value; }
+
+    @Override
+    Object copy(Object array, int arraySize, int sourceFrom, int destinationFrom, int size) {
         if (size == 0) {
             return new Object[arraySize];
         } else {
-            final Object result = new Object[arraySize];
-            System.arraycopy(source, sourceFrom, result, destinationFrom, size);
+            final Object[] result = new Object[arraySize];
+            System.arraycopy(array, sourceFrom, result, destinationFrom, size); // has to be near the object allocation to avoid zeroing out the array
             return result;
         }
     }
+}
 
-    /** array access, avoiding call-site casts */
-    @SuppressWarnings("unchecked")
-    static <T> T getAt(Object array, int index) { return (T) ((Object[]) array)[index]; }
-    @SuppressWarnings("unchecked")
-    static void setAt(Object array, int index, Object value) { ((Object[]) array)[index] = value; }
-    static int lengthOf(Object array) {
-        return ((Object[]) array).length;
+final class IntArray extends LeafType<Integer> {
+    private static final long serialVersionUID = 1L;
+    static final IntArray INT = new IntArray();
+
+    @Override
+    Class<Integer> type() { return int.class; }
+
+    private static final int[] EMPTY = new int[0];
+
+    @Override
+    int[] empty() { return EMPTY; }
+
+    private static int[] cast(Object array) { return (array == null) ? EMPTY : (int[]) array; }
+
+    @Override
+    int lengthOf(Object array) { return cast(array).length; }
+
+    @Override
+    Integer getAt(Object array, int index) { return cast(array)[index]; }
+
+    @Override
+    void setAt(Object array, int index, Object value) { cast(array)[index] = (Integer) value; }
+
+    @Override
+    Object copy(Object array, int arraySize, int sourceFrom, int destinationFrom, int size) {
+        if (size == 0) {
+            return new int[arraySize];
+        } else {
+            final int[] result = new int[arraySize];
+            System.arraycopy(array, sourceFrom, result, destinationFrom, size); // has to be near the object allocation to avoid zeroing out the array
+            return result;
+        }
+    }
+}
+
+@SuppressWarnings("unchecked")
+@GwtIncompatible
+final class PrimitiveArray<T> extends LeafType<T> {
```

Comment:
generic converter for non-optimized array access (only needed if we don't privide converters for every primitive)